src/connection.cpp

```diff
@@ -180,7 +185,12 @@ static int connSocketRead(connection *conn, void *buf, size_t buf_len) {
         conn->state.store(CONN_STATE_CLOSED, std::memory_order_release);
     } else if (ret < 0 && errno != EAGAIN) {
         conn->last_errno = errno;
-        conn->state.store(CONN_STATE_ERROR, std::memory_order_release);
+
+        /* Don't overwrite the state of a connection that is not already
+         * connected, not to mess with handler callbacks.
+         */
+        if (conn->state == CONN_STATE_CONNECTED)
```

Comment:
You need to use compare exchange here because the value of conn->state can change between the if statement and the actual value change below.  That's why you need an atomic compare exchange here.